### ISSUE TITLE
Getting instance name from google_compute_instance datasource

### ISSUE THREAD

**Issue by nosammai** [hashicorp/terraform-provider-google#12112]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Description
I am trying to create DNS entries for all of the instances created by a managed instance group, but get an error when attempting to grab the instance names from the google_compute_instance datasource.

I was able to work around this by grabbing the instance name off the end of the instance urls returned by the google_compute_managed_instance_group resource, but that doesn't really seem ideal.

### Terraform Version
```
Terraform v1.1.9
on linux_amd64
+ provider registry.terraform.io/hashicorp/google v4.27.0
```

### Affected Resource(s)

* google_compute_instance datasource

### Terraform Configuration Files

```tf
locals {
  instance_names = data.google_compute_instance.instance[*].name
}

data "google_compute_region_instance_group" "mig" {
  self_link = google_compute_region_instance_group_manager.mig.instance_group
}

data "google_compute_instance" "instance" {
  count     = var.instance_count
  self_link = data.google_compute_region_instance_group.mig.instances[count.index].instance
}

resource "google_compute_region_instance_group_manager" "mig" {
  name = "${module.template.base_hostname}-mig"
  target_size                      = var.instance_count
  wait_for_instances               = true
<omitted a bunch of arguments for brevity since it's not relevant>
}

resource "google_dns_record_set" "instance_dns" {
  count        = var.instance_count
  name         = "${local.instance_names[count.index]}.${data.google_dns_managed_zone.project_dns_zone.dns_name}"
  project      = var.gcp_subnet_project
  type         = "A"
  ttl          = 300
  managed_zone = data.google_dns_managed_zone.project_dns_zone.name
  rrdatas      = [data.google_compute_instance.instance[count.index].network_interface[0].network_ip]
}
```

Workaround:
```tf
locals {
  instance_names = [
    for instance in data.google_compute_region_instance_group.mig.instances[*].instance :
    split("/", instance)[length(split("/", instance)) - 1]
  ]
}
```

### Expected Behavior

`name` attribute would be looked up at apply-time and templated into the string

### Actual Behavior

Error at plan-time saying that `name` is null (presumably because it is an argument that I did not set, because I opted to use self_link)

```
╷
│ Error: Invalid template interpolation value
│
│   on ../../../compute/gcp/linux_mig/main.tf line 80, in resource "google_dns_record_set" "instance_dns":
│   80:   name         = "${local.instance_names[count.index]}.${data.google_dns_managed_zone.project_dns_zone.dns_name}"
│     ├────────────────
│     │ count.index is 0
│     │ local.instance_names is tuple with 2 elements
│
│ The expression result is null. Cannot include a null value in a string
│ template.
```

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

If I put in a dummy value for the DNS entry and apply that, then swap the config back to using the `name` attribute from the datasource, it will work the 2nd time, presumably because it has already gone and fetched the data/filled in the name previously and knows it isn't null.

**Comment by edwardmedia** [comment]
@nosammai can you share what you see if you run `terraform show`? Particularly the data source `google_compute_region_instance_group`?

**Comment by edwardmedia** [comment]
@nosammai can you try function tolist() like below to see if that works for you?

```
rrdatas  = tolist(data.google_compute_instance.instance)[count.index].network_interface[0].network_ip
```

**Comment by edwardmedia** [comment]
@nosammai don't you get anything from below outputs?

```
output "instance_ip_addr" {
  value = tolist(data.google_compute_instance.instance)[0].network_interface[0].network_ip
}

output "instance" {
  value = tolist(data.google_compute_instance.instance)[0].name
}
```

**Comment by edwardmedia** [comment]
@nosammai how did you get [this](https://github.com/hashicorp/terraform-provider-google/issues/12112#issuecomment-1188353387) when the plan is failing? You may want to comment the issue resources first to debug the problem....

**Comment by edwardmedia** [comment]
@nosammai  did you run `apply`?

**Comment by edwardmedia** [comment]
@nosammai can you try to test with a single node (just index = 0)? If not working, can you post your config and the debug log?

**Comment by edwardmedia** [comment]
@nosammai it is failing on line 82 on `name` field. Looks like on your `local.instance_name[x]`. How come it is related to the state?

```tf
resource "google_dns_record_set" "instance_dns" {
  count        = var.instance_count
  name         = "${local.instance_names[count.index]}.${data.google_dns_managed_zone.project_dns_zone.dns_name}" <-----
  project      = var.gcp_subnet_project
  type         = "A"
  ttl          = 300
  managed_zone = data.google_dns_managed_zone.project_dns_zone.name
  rrdatas      = [tolist(data.google_compute_instance.instance)[count.index].network_interface[0].network_ip]
}
```

**Comment by edwardmedia** [comment]
@nosammai do you still have an issue with using the data source? Try below line. It works for me. I did not see it was failing in your [error message](https://github.com/hashicorp/terraform-provider-google/issues/12112#issuecomment-1198328575). 

```
rrdatas      = [tolist(data.google_compute_instance.instance)[count.index].network_interface[0].network_ip]
```

**Comment by edwardmedia** [comment]
@nosammai how about this?

```tf
resource "google_dns_record_set" "instance_dns" {
  count        = 3
  name         = "${tolist(data.google_compute_instance.instance)[count.index].name}.${google_dns_managed_zone.project_dns_zone.dns_name}"
  type         = "A"
  ttl          = 300
  managed_zone = google_dns_managed_zone.project_dns_zone.name
  rrdatas      = [tolist(data.google_compute_instance.instance)[count.index].network_interface[0].network_ip]
}
```

**Comment by slevenick** [comment]
This sort of makes sense to me based on the weird way the data source looks up the instance itself. It's a super old resource so it behaves a little strangely I think.

Basically, *most* datasources and resources require fields like name, location, zone to be set, but it looks like this datasource allows you to identify it by its self_link which is what you are doing. This work fine, but results in the `name` not being marked as computed, which is normally what we want for datasources but doesn't work in this case where the instance is identified by self_link. This all happens in the `dataSourceGoogleComputeInstance` function

Two possible solutions here:

1. Identify the datasource in your config with the `name` and `zone` fields rather than self_link
2. Calculate the `name` from the `self_link` or `id` of the datasource

Example of 1:
```hcl
data "google_compute_instance" "instance" {
  count     = var.instance_count
  name = element(split("/", data.google_compute_region_instance_group.mig.instances[count.index].instance), length(split("/", data.google_compute_region_instance_group.mig.instances[count.index].instance))-1)
}
```
This splits the self_link on "/" and grabs the last element. Now that `name` is set in the resource it can be used to compute the other variable

You have something similar to example 2 in your workaround in the original issue

We *could* fix the issue by marking this field as Computed, but that's not easy with our normal datasource generation tools. It could be done by hand though

**Comment by slevenick** [comment]
If the format of the self_link changed the provider code would have a bad time. I think it's fairly safe to depend on the name being the last element after splitting on "/"

**Comment by slevenick** [comment]
Not that I know of. If this was implemented today I would lean towards not supporting identifying the instance by self_link, and marking `name` as required

**Comment by slevenick** [comment]
I don't believe we actually receive this information from the API. If we implemented it we would simply be doing the same split on "/" and grab the last element logic.

I believe that's the best way to get just the name. Honestly the mix of names vs URLs in some of these Compute resources is frustrating, and there's not a great way to handle them
